src/isomorphic/classic/element/ReactElementValidator.js

```diff
@@ -181,6 +181,31 @@ function validatePropTypes(element) {
   }
 }
 
+/**
+ * Given an element, freeze `type` and `defaultProps` (Only used in DEV)
+ *
+ * @param {ReactElement} element
+ */
+function freezeTypeAndDefaultProps(element) {
+  var type = element.type;
+  if (typeof type !== 'function' || type.wasTaggedReactComponent) {
+    return;
+  }
+  if (canDefineProperty) {
+    Object.defineProperty(type, 'wasTaggedReactComponent', { value: true });
+    Object.defineProperty(type, 'defaultProps', {
+      writable: false,
+      configurable: false,
+    });
+  } else {
+    type.wasTaggedReactComponent = true;
+  }
+  if (Object.freeze) {
+    Object.freeze(type);
```

Comment:
This looks good to me. Although after thinking a bit more about it, freezing the entire `.type` object might be too restrictive.

We only really need to freeze the default props themselves and the property itself to make this inlineable.

Maybe we can just drop this line?